### PR TITLE
pkg/prometheus/ha: store and hash secrets properly

### LINKED ISSUE #62: Scraping Service: `basic_auth` gets wiped by marshaling to YAML
Prometheus stores the `basic_auth` as a `Secret` type which implements `yaml.Marshaler` and always forces the string to be `<secret>`. This breaks the configuration storage.

The `yamlCodec` in `pkg/prometheus/ha` should wrap the `instance.Config` type and store secret values separately. The following are secrets:

- BasicAuth Password
- HTTP Client bearer token

Note that for the same reason, hashing configs when detecting if configs changed is broken; the value being hashed is not the secret but rather the string `<secret>`. To fix this, the secret values above should be added separately to the hashing function if they are non-nil and non-empty.

### PR DESCRIPTION
**Pull request by rfratto**
Fixes #62.

/cc @gotjosh, @hoenn: no rush on getting this merged, waiting until Monday is fine

### CODE REVIEW

File: pkg/prometheus/ha/codec.go [hunk 1]
```diff
@@ -60,3 +68,45 @@ func (*yamlCodec) Encode(v interface{}) ([]byte, error) {
...
+}
+
+// newCodecConfig creates a new codecConfig that is ready for marshaling as YAML
+// and storing in a KV store. newCodecConfig must extract secrets stored in the
+// instance.Config and store them elsewhere as strings as the MarshalYAML function
+// on the Prometheus Secret type replaces the contents of the secret with
+// the text "<secret>".
+func newCodecConfig(c *instance.Config) *codecConfig {
+	var secrets []configSecrets
+	for _, rwr := range c.RemoteWrite {
```
Comment (gotjosh) [inline pkg/prometheus/ha/codec.go:84]: is order when marshalling/unmarshalling guaranteed to be preserved?

Comment (gotjosh) [reply pkg/prometheus/ha/codec.go:84]: Also, this means that if you move sections in the file it is now broken? Maybe this is isn't a problem if we don't support it anyways.

Comment (rfratto) [reply pkg/prometheus/ha/codec.go:84]: Yep, the [JSON spec](https://www.rfc-editor.org/rfc/rfc7159.txt) says that arrays are ordered - this seems to hold in our [YAML decoder](https://github.com/go-yaml/yaml/blob/0b1645d91e851e735d3e23330303ce81f70adbe3/decode.go#L201) as well.

As for moving sections, we require users to always upload the whole config file each time (including the secrets). So if you move the sections around in the file and update the config in the API, `newCodecConfig` with the write ordering would completely supplant the old one. 

It's possible someone could modify the KV store directly and mess up our ordering, but there's probably nothing we can (or should?) do about that :)

Comment (gotjosh) [reply pkg/prometheus/ha/codec.go:84]: Makes perfect sense.

**Pull request merged**